app/datatables/application_datatable.rb

```diff
@@ -0,0 +1,88 @@
+class ApplicationDatatable
+  attr_reader :base_relation, :params
+
+  def initialize(base_relation, params)
+    @base_relation = base_relation
+    @params = params
+  end
+
+  def as_json(*)
+    {
+      data: sanitize(data),
+      recordsFiltered: filtered_records.size,
+      recordsTotal: base_relation.size
+    }
+  end
+
+  private
+
+  def sanitize(data)
+    if data.is_a? Array
+      data.map { |datum| sanitize datum }
+    elsif data.is_a? Hash
+      data.transform_values! { |value| sanitize value }
+    else
+      ERB::Util.html_escape data
+    end
+  end
+
+  def filtered_records
+    raw_records
+  end
+
+  def paginated_records
+    filtered_records
+      .offset(offset)
+      .limit(limit)
+  end
+  alias_method :records, :paginated_records
+
+  def additional_filters
+    @additional_filters ||= params[:additional_filters] || {}
+  end
+
+  def search_term
+    @search_term ||= params[:search][:value]
+  end
+
+  def build_order_clause
+    Arel.sql "#{order_by} #{order_direction}" if order_by.present?
```

Comment:
@compwron I try to avoid it, since **A**rel **R**eally **E**xasperates **L**ogicians, but alas...
```
DEPRECATION WARNING: Dangerous query method (method whose arguments are used as raw SQL) called
with non-attribute argument(s): "COALESCE(users.display_name, users.email) ASC". Non-attribute arguments
will be disallowed in Rails 6.1. This method should not be called with user-provided values, such as
request parameters or model attributes. Known-safe values can be passed by wrapping them in Arel.sql()
```